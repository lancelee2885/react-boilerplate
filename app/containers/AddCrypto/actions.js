/*
 *
 * AddCrypto actions
 *
 */

import {
  CHANGE_FORM_DATA,
  SUBMIT_FORM_DATA,
  SUBMIT_FORM_DATA_ERROR,
  SUBMIT_FORM_DATA_SUCCESS,
} from './constants';

/**
 * Changes the input fields of the form
 */
export function changeFormData(formData) {
  return {
    type: CHANGE_FORM_DATA,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function submitFormData() {
  return {
    type: SUBMIT_FORM_DATA,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataSubmitted(formData) {
  return {
    type: SUBMIT_FORM_DATA_SUCCESS,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataError(err) {
  return {
    type: SUBMIT_FORM_DATA_ERROR,
    err,
  };
}
