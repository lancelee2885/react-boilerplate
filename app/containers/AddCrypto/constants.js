/*
 *
 * AddCrypto constants
 *
 */

export const CHANGE_FORM_DATA = 'app/AddCrypto/CHANGE_FORM_DATA';
export const SUBMIT_FORM_DATA = 'app/AddCrypto/SUBMIT_FORM_DATA';
export const SUBMIT_FORM_DATA_SUCCESS =
  'app/AddCrypto/SUBMIT_FORM_DATA_SUCCESS';
