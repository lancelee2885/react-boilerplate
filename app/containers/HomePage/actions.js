/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_CRYPTO,
  LOAD_CRYPTO_ERROR,
  LOAD_CRYPTO_SUCCESS,
} from './constants';

/**
 * Load the crypto from backend, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_CRYPTO
 */
export function loadCryptos() {
  return {
    type: LOAD_CRYPTO,
  };
}

/**
 * Dispatched when the cryptocurrencies are loaded by the request saga
 *
 * @param  {array} cryptos The crypto data
 *
 * @return {object}      An action object with a type of LOAD_CRYPTO_SUCCESS passing the cryptos
 */
export function cryptosLoaded(cryptos) {
  return {
    type: LOAD_CRYPTO_SUCCESS,
    cryptos,
  };
}

/**
 * Dispatched when loading the cryptocurrencies fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_CRYPTO_ERROR passing the error
 */
export function cryptossLoadingError(error) {
  return {
    type: LOAD_CRYPTO_ERROR,
    error,
  };
}
