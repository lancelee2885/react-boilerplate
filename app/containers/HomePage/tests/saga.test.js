/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_CRYPTO } from '../constants';
import { cryptosLoaded, cryptossLoadingError } from '../actions';

import loadCryptos, { fetchCryptos } from '../saga';

const mockCryptos = ['test'];

/* eslint-disable redux-saga/yield-effects */
describe('fetchCryptos Saga', () => {
  let getCryptosGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getCryptosGenerator = fetchCryptos();

    const callDescriptor = getCryptosGenerator.next(mockCryptos).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the cryptosLoaded action if it requests the data successfully', () => {
    const response = { cryptocurrency: mockCryptos };
    const putDescriptor = getCryptosGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(cryptosLoaded(mockCryptos)));
  });

  it('should call the cryptossLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getCryptosGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(cryptossLoadingError(response)));
  });
});

describe('loadCryptos Saga', () => {
  const githubDataSaga = loadCryptos();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_CRYPTO, fetchCryptos));
  });
});
