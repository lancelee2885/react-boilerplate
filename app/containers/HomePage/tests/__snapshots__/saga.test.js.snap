// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCryptos Saga should call the cryptossLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`fetchCryptos Saga should dispatch the cryptosLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3000/api/",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
